### PR TITLE
Request For Feedback: DO NOT MERGE: Adding TC: v0 -- Work In progress

### PR DESCRIPTION
**Pull request by mimullin-bbry**
@danobi @anakryiko 

Request for Feedback.

Adding BPF Tc into libbpf-rs.

Please give guidance on
1) Whether the re-exposing of BPF_TC_INGRESS/EGRESS/etc is acceptable

2) the builder-like pattern used to generate a TcHook (see the example for how to build a TcHook in two different ways)

3) How to properly link to struct Program in program.rs in rust-doc, and how to link to the example program (examples/tc_whitelist_ports/src/main.rs) in rust-dock

Thank you for your time.  This is a 'spare time' piece of code where I am using libbpf-rs to learn about Traffic Control.

### CODE REVIEW

**Comment by insearchoflosttime** [comment]
Hi @mimullin-bbry,

1 - The way you expose BPF_TC_INGRESS/EGRESS/CUSTOM looks fine to me.

2 - Either way you're using the TC hook looks fine to me. 

3 - I think you're linking to it correctly in program.rs for rust-doc. As for linking to the example program, the only example we currently mention is runqslower (https://github.com/libbpf/libbpf-rs/blob/81417afc7b57610e4cea2c1043cf0ee4a6fc5fc8/libbpf-rs/src/lib.rs#L63-L67)

File: libbpf-rs/src/tc.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+/// This object exposes operations to create and attach a bpf_tc_hook
+/// into the TC subsystem
+#[derive(Clone, Copy)]
+pub struct TcHook {
+    hook: libbpf_sys::bpf_tc_hook,
+    opts: libbpf_sys::bpf_tc_opts,
+}
+
+impl TcHook {
+    /// Create a New TcHook given the file descriptor of the loaded SEC("classifier") Program
```
Comment (anakryiko) [inline libbpf-rs/src/tc.rs:31]: s/classifier/tc/, it's a better and more modern naming

File: libbpf-rs/src/tc.rs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+///
+/// This object exposes operations to create and attach a bpf_tc_hook
+/// into the TC subsystem
+#[derive(Clone, Copy)]
+pub struct TcHook {
+    hook: libbpf_sys::bpf_tc_hook,
+    opts: libbpf_sys::bpf_tc_opts,
+}
+
+impl TcHook {
```
Comment (anakryiko) [inline libbpf-rs/src/tc.rs:30]: TcHook actually mixes builder and actual hook object into one object. Look at other objects in libbpf-rs (e.g., Object and ObjectBuilder), we should probably follow that pattern.

**Review by anakryiko — CHANGES_REQUESTED** [review]
We probably shouldn't add a 3rd copy of 2MB vmlinux.h. Can you please make sure that we only have one copy and that copy is used by every example. No need to download megabytes of duplicated header when checking out libbpf-rs repo.

**Comment by mimullin-bbry** [comment]
> We probably shouldn't add a 3rd copy of 2MB vmlinux.h. Can you please make sure that we only have one copy and that copy is used by every example. No need to download megabytes of duplicated header when checking out libbpf-rs repo.

Yes, but I'll do that work under a separate ticket (https://github.com/libbpf/libbpf-rs/issues/143).  No need to confuse the other examples in a PR specific to TC.

File: libbpf-rs/src/tc.rs (hunk 2 above)
Comment (mimullin-bbry) [reply libbpf-rs/src/tc.rs:30]: +1

File: libbpf-rs/src/tc.rs (hunk 1 above)
Comment (mimullin-bbry) [reply libbpf-rs/src/tc.rs:31]: +1

File: libbpf-rs/src/program.rs [hunk 3]
```diff
@@ -344,4 +344,10 @@ impl Program {
             Ok(Link::new(ptr))
         }
     }
+
+    /// Get a TcHook in order to attach a BPF TC Program [BPF
+    /// TC](https://lwn.net/ml/bpf/20210512103451.989420-3-memxor@gmail.com/)
+    pub fn get_tc_hook(&self) -> TcHook {
```
[the remark this answers is not in the document]
Comment (mimullin-bbry) [reply libbpf-rs/src/program.rs:350]: sounds good.

File: libbpf-rs/src/tc.rs (hunk 1 above)
Comment (mimullin-bbry) [reply libbpf-rs/src/tc.rs:37]: The version of libbpf used by libbpf-sys does not have the necessary changes for using SEC("tc") yet.

**Comment by mimullin-bbry** [comment]
Im fairly happy with this change now.  Pending the changes to https://github.com/libbpf/libbpf-rs/pull/149 I'll make some tweaks based on using only a single vmlinux.h file and remove this from "Draft"

**Pull request closed without merging**